src/content/1.7/modules/concepts/controllers/admin-controllers/override-decorate-controller.md

```diff
@@ -0,0 +1,153 @@
+{{% /notice %}}
+
+## Decorate the controller
+
+However overriding a controller or a service is usually not recommanded, because this means you completely rewrite its behavior. If the implementation is updated in later versions of PrestaShop, your override will ignore these updates, which might create bugs.
+
+It is better to use [decoration](https://symfony.com/doc/3.4/service_container/service_decoration.html) which keeps the behavior of the decorated controller while allowing you to extend its behavior by modiying received inputs and outputs.
+
+Back to the controller responsible for the page "Improve > Design > CMS Pages" which is registered with service ID `PrestaShopBundle\Controller\Admin\Improve\Design\CmsPageController`.
+
+With the following configuration item, we can decorate it with a custom controller in our module:
+```yaml
+# modules/your-module/config/services.yml
+  custom_controller:
+    class: MyModule\Controller\DemoController
+    decorates: PrestaShopBundle\Controller\Admin\Improve\Design\CmsPageController
+    arguments: ['@custom_controller.inner']
+```
+
+Thanks to this, whenever Symfony forwards a request to the Core controller `PrestaShopBundle\Controller\Admin\Improve\Design\CmsPageController` it will be forwarded to `DemoController` instead. But what is different with overriding is that the decorated CmsPageController is injected into the constructor of DemoController, and we can use it:
+```php
+<?php
+
+namespace MyModule\Controller;
+
+use PrestaShop\PrestaShop\Core\Search\Filters\CmsPageCategoryFilters;
+use PrestaShop\PrestaShop\Core\Search\Filters\CmsPageFilters;
+use PrestaShopBundle\Controller\Admin\FrameworkBundleAdminController;
+use PrestaShopBundle\Controller\Admin\Improve\Design\CmsPageController;
+use Symfony\Component\HttpFoundation\Request;
+
+class DemoController extends FrameworkBundleAdminController
+{
+    /**
+     * @var CmsPageController
+     */
+    private $decoratedController;
+
+    public function __construct(CmsPageController $decoratedController)
+    {
+        $this->decoratedController = $decoratedController;
+    }
+
+    public function indexAction(CmsPageCategoryFilters $categoryFilters, CmsPageFilters $cmsFilters, Request $request)
+    {
+        return $this->decoratedController->indexAction($categoryFilters, $cmsFilters, $request);
+    }
+}
+```
+In this example I do nothing else than returning the exact output from the decorated controller.
+However I could modify the input request or I could modify the output given by decorated controller before returning it. Example:
+```php
+    public function indexAction(CmsPageCategoryFilters $categoryFilters, CmsPageFilters $cmsFilters, Request $request)
+    {
+        $output = $this->decoratedController->indexAction($categoryFilters, $cmsFilters, $request);
+
+        $myService = $this->getMyPaymentService();
+        $output = $this->injectMyData($myService, $output);
+
+        return $output;
+    }
+```
+
+
```

Comment:
There's a missing `__call()` function to account for other actions that wouldn't have been decorated by the controller.

Reply:
Cannot be done. When I insert the `__call()` I have multiple symfony event listeners throwing ReflectionException

Needs to be explored and fixed ... outside of the scope of this PR 😭